idd/Energy+.idd.in

```diff
@@ -1277,6 +1277,7 @@ RunPeriod,
        \memo Multiple run periods may be input, but they may not overlap.
        \min-fields 7
   A1 , \field Name
+       \required-field
```

Comment:
So I tried to run an example file with a blank run period name, expecting it to be aborted by input processor validation, but it actually succeeded.  Here is the output...

```
$ git status
On branch PR_opened/Fix_6937_RequireRunPeriodName

$ grep -A 3 'RunPeriod,' 1ZoneUncontrolled.idf 
  RunPeriod,
    ,                        !- Name
    1,                       !- Begin Month
    1,                       !- Begin Day of Month

$ ./energyplus -w /eplus/epw/golden.epw 1ZoneUncontrolled.idf 
EnergyPlus Starting
EnergyPlus, Version 9.1.0-c4dea36293, YMD=2019.02.14 09:02
Adjusting Air System Sizing
Adjusting Standard 62.1 Ventilation Sizing
Initializing Simulation
Reporting Surfaces
Beginning Primary Simulation
Initializing New Environment Parameters
Warming up {1}
  (snip)
Warming up {22}
Starting Simulation at 12/21 for DENVER CENTENNIAL  GOLDEN   N ANN HTG 99% CONDNS DB
Initializing New Environment Parameters
Warming up {1}
  (snip)
Warming up {16}
Starting Simulation at 07/21 for DENVER CENTENNIAL  GOLDEN   N ANN CLG 1% CONDNS DB=>MWB
Initializing New Environment Parameters
Warming up {1}
  (snip)
Warming up {20}
Starting Simulation at 01/01/2013 for RUNPERIOD 1
Updating Shadowing Calculations, Start Date=01/21/2013
  (snip)
Continuing Simulation at 12/07/2013 for RUNPERIOD 1
Updating Shadowing Calculations, Start Date=12/27/2013
Continuing Simulation at 12/27/2013 for RUNPERIOD 1
  (snip)
EnergyPlus Completed Successfully.
```

So somewhere in the code it is still allowing a blank and applying a numeric counter to the run period object.  I was also going to comment on the `\note` a few lines down where it says it can be entered blank and the code will use a weather file name.  This seems unfinished at this point.

@mjwitte or @jasondegraw do you have any thoughts on this?  I don't mind trying to push a little bit on it, but I was hoping it would be a straightforward change to review.

Reply:
@Myoldmopar It is very likely due to [this bit of code.](https://github.com/NREL/EnergyPlus/blob/1dc4ef01450b86ac85ef5c98e4055de65ee06788/src/EnergyPlus/InputProcessing/IdfParser.cc#L266-L268)

Reply:
Thanks @mbadams5, I can't immediately tell how much of that code should be removed to make the input processor fail when the value is blank?  Kinda looks like more than those three lines alone, but maybe it is?

Reply:
@mbadams5 @jasondegraw taking out that input processor did not solve the problem.  The input processor is still somehow letting the IDF through.  But now, instead of applying the index in the input processor and reporting it as "RunPeriod 1", it shows:

```
Starting Simulation at 01/01/2013 for Denver Centennial  Golden   Nr CO USA TMY3 WMO#=724666
Updating Shadowing Calculations, Start Date=01/21/2013
Continuing Simulation at 01/21/2013 for Denver Centennial  Golden   Nr CO USA TMY3 WMO#=724666
Updating Shadowing Calculations, Start Date=02/10/2013
```

I found the code that reads in runperiod data, and as expected, it ultimately just allows the title to be read in as-is, including blanks, because it was counting on the input processor to give it an indexed name if the user left it blank.  I considered adding in a FatalError there to require the user to input a value, but I'd prefer to understand why the IP is even letting the value through first.  

I'm also feeling inclined to pushing this until later, like post release.  We should probably have a chance to discuss this because this _could_ affect all the interfaces if they weren't writing out run period names.  I'm not opposed to making them required, it's just starting to feel like this needs more thought.

Thoughts @mjwitte @jasondegraw @mbadams5 @jmarrec ?

Reply:
Interesting. @Myoldmopar Have you tried leaving a required field blank in some other object? Maybe that's not getting enforced at all?

Reply:
@Myoldmopar I say let's kick this to the next release just to be safe. I do believe that it is a needed change, but I 100% agree that we need to better understand how things are getting through.

Reply:
I second the motion to punt for today.  It's definitely not a major pain point, but will be nice to clean up with some time.  Moving to the future release.

Reply:
> Interesting. @Myoldmopar Have you tried leaving a required field blank in some other object? Maybe that's not getting enforced at all?

@mjwitte I did try this by leaving the name blank on the design day, it fails instantly as expected, with:

```
Program Version,EnergyPlus, Version 9.2.0-8e50ff5c50, YMD=2019.05.28 13:01,
   ** Severe  ** SizingPeriod:DesignDay Name, cannot be blank
   **  Fatal  ** GetWeatherInput: Above errors cause termination
   ...Summary of Errors that led to program termination:
```

It's definitely specific to the run period.

Reply:
@mbadams5 @jmarrec @mjwitte I re-confirmed the above, that cleaning up that special run-period processing block does **not** solve the problem.  It runs fine with a blank name still.  I went ahead and pushed that code up though, along with the branch being merged up with develop.  I'm curious if it causes any issues when CI runs everything.

As far as I can tell though, this is **not** yet fixed, so we need to figure out why blank run period is still getting through.  If someone wants to confirm my status on this branch with blank run period name (passes) and blank DD name (fails), that would be reassuring.

Reply:
I will do that.

Reply:
That's partly because you missed another line above (L265) where it already does `std::string name = obj_name + " " + s;`.
If you do that then you'll likely end up with a crash (at least I do with `1ZoneUncontrolled.idf`) `** Severe  ** Duplicate name found. name: "". Overwriting existing object.` but that's still not great because it's not explicit.

In fact, what's needed is for the `IdfParser.cc` is to check the schema for name `is_required` which is done nowhere currently: if name is blank, but schema says name has `"is_required": true`, then issue a crash saying name cannot be blank. I will do that and push.

epJSON extract
```
        "RunPeriod": {
            "name": {
                "note": "descriptive name (used in reporting mainly) If blank, weather file title is used.  If not blank, must be unique", 
                "is_required": true, 
                "type": "string", 
                "reference": [
                    "RunPeriodsAndDesignDays"
                ]
            }, 
```

Reply:
That sounds right, thanks for finding that and pushing on it.  Just to be clear, in our lingo, a severe/fatal is not a crash, that's just a properly handled fatal error.  A crash would be a divide by zero, unhandled exception.

Reply:
I misused "crash" above, I meant fatal error indeed.